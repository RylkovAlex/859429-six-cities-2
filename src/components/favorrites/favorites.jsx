import React from 'react';
import PropTypes from 'prop-types';
import {Link} from 'react-router-dom';

import {connect} from 'react-redux';
import ActionCreator, {Operation} from '../../redux/actions/action-creator/action-creator';
import {getFavorites, getCities} from '../../redux/selectors/selectors';

import {firstToUpperCase} from '../../utils/utils';
import Header from '../header/header.jsx';
import FavoritesEmpty from '../favorites-empty/favorites-empty.jsx';
import OffersList, {ListType} from '../offers-list/offers-list.jsx';
import withAuth from '../../hocs/with-auth/with-auth.jsx';

const Favorites = ({favorites, postFavorite, activeCity, cities, setCity}) => {
  const handleCityLinkClick = (cityName) => {
    const city = cities.find((c) => c.name === cityName);
    setCity(city);
  };
  if (favorites.every((group) => !group.hotels.length)) {
    return <FavoritesEmpty/>;
  }
  return (
    <div className="page">
      <Header />
      <main className="page__main page__main--favorites">
        <div className="page__favorites-container container">
          <section className="favorites">
            <h1 className="favorites__title">Saved listing</h1>
            <ul className="favorites__list">
              {favorites.map((group) => {
                if (!group.hotels.length) {
                  return null;
                }
                return (
                  <li key={group.city} className="favorites__locations-items">
                    <div className={`favorites__locations locations ${group.city === activeCity.name ? `locations--current` : ``}`}>
                      <div className="locations__item">
                        <Link to="/"
                          className="locations__item-link"
                          href="#"
                          onClick={() => handleCityLinkClick(group.city)}
                        >
                          <span>{firstToUpperCase(group.city)}</span>
                        </Link>
                      </div>
                    </div>
                    <OffersList
                      offerCards = {group.hotels}
                      listType = {ListType.FavoriteList}
                      handleCardHover = {()=>{}}
                      handleBookmarkClick = {postFavorite}
                    />
                  </li>
                );
              })}
            </ul>
          </section>
        </div>
      </main>
      <footer className="footer container">
        <a className="footer__logo-link" href="main.html">
          <img
            className="footer__logo"
            src="img/logo.svg"
            alt="6 cities logo"
            width="64"
            height="33"
          />
        </a>
      </footer>
    </div>
  );
};

Favorites.propTypes = {
  favorites: PropTypes.array.isRequired,
  postFavorite: PropTypes.func.isRequired,
  activeCity: PropTypes.object.isRequired,
  cities: PropTypes.array.isRequired,
  setCity: PropTypes.func.isRequired,
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  activeCity: state.city,
  cities: getCities(state),
  favorites: getFavorites(state),
});

const mapDispatchToProps = (dispatch) => ({
  setCity: (city) => dispatch(ActionCreator.changeCity(city)),
  postFavorite: (cardId, status) => dispatch(Operation.postFavorite(cardId, status)),
});

export {Favorites};
export default withAuth(connect(mapStateToProps, mapDispatchToProps)(Favorites));
