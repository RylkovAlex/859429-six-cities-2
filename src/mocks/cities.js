export const cityMock = {
  name: `Amsterdam`,
  location: {
    latitude: 52.370216,
    longitude: 4.895168,
    zoom: 10,
  },
};

export const citiesMock = [
  {
    name: `1`,
    location: {
      latitude: 1,
      longitude: 2,
      zoom: 3,
    },
  },
  {
    name: `2`,
    location: {
      latitude: 1,
      longitude: 2,
      zoom: 3,
    },
  },
];
