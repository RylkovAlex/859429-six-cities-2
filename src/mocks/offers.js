export const offerCards = ([
  {
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10
      }
    },
    id: 1,
    previewImage: `img/apartment-01.jpg`,
    images: [`../img/room.jpg`, `../img/apartment-01.jpg`, `../img/apartment-02.jpg`, `../img/apartment-03.jpg`, `../img/studio-01.jpg`, `../img/apartment-01.jpg`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: false,
    isPremium: false,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina`,
      avatarUrl: `../img/avatar-angelina.jpg`,
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 52.3909553943508,
      longitude: 4.85309666406198,
      zoom: 8
    }
  },
  {
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10
      }
    },
    id: 2,
    previewImage: `img/apartment-02.jpg`,
    images: [`../img/room.jpg`, `../img/apartment-01.jpg`, `../img/apartment-02.jpg`, `../img/apartment-03.jpg`, `../img/studio-01.jpg`, `../img/apartment-01.jpg`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: false,
    isPremium: true,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Wi-Fi`, `Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina2`,
      avatarUrl: `../img/avatar-angelina.jpg`,
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 52.369553943508,
      longitude: 4.85309666406198,
      zoom: 8
    }
  },
  {
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10
      }
    },
    id: 3,
    previewImage: `img/apartment-03.jpg`,
    images: [`../img/room.jpg`, `../img/apartment-01.jpg`, `../img/apartment-02.jpg`, `../img/apartment-03.jpg`, `../img/studio-01.jpg`, `../img/apartment-01.jpg`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: true,
    isPremium: true,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina3`,
      avatarUrl: `../img/avatar-angelina.jpg`,
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 52.3909553943508,
      longitude: 4.929309666406198,
      zoom: 8
    }
  },
  {
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10
      }
    },
    id: 4,
    previewImage: `img/apartment-01.jpg`,
    images: [`../img/room.jpg`, `../img/apartment-01.jpg`, `../img/apartment-02.jpg`, `../img/apartment-03.jpg`, `../img/studio-01.jpg`, `../img/apartment-01.jpg`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: false,
    isPremium: true,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina4`,
      avatarUrl: `../img/avatar-angelina.jpg`,
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 52.3809553943508,
      longitude: 4.939309666406198,
      zoom: 8
    },
  },
  {
    city: {
      name: `Cologne`,
      location: {
        latitude: 50.938903,
        longitude: 6.966074,
        zoom: 10
      }
    },
    id: 5,
    previewImage: `img/apartment-01.jpg`,
    images: [`../img/room.jpg`, `../img/apartment-01.jpg`, `../img/apartment-02.jpg`, `../img/apartment-03.jpg`, `../img/studio-01.jpg`, `../img/apartment-01.jpg`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: false,
    isPremium: true,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina4`,
      avatarUrl: `../img/avatar-angelina.jpg`,
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 50.938903,
      longitude: 6.966074,
      zoom: 8
    },
  },
  {
    city: {
      name: `Paris`,
      location: {
        latitude: 48.851353,
        longitude: 2.341412,
        zoom: 10
      }
    },
    id: 6,
    previewImage: `img/apartment-01.jpg`,
    images: [`../img/room.jpg`, `../img/apartment-01.jpg`, `../img/apartment-02.jpg`, `../img/apartment-03.jpg`, `../img/studio-01.jpg`, `../img/apartment-01.jpg`],
    title: `Beautiful & luxurious studio at great location`,
    isFavorite: false,
    isPremium: true,
    rating: 4.8,
    type: `apartment`,
    bedrooms: 3,
    maxAdults: 4,
    price: 120,
    goods: [`Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    host: {
      id: 3,
      isPro: true,
      name: `Angelina4`,
      avatarUrl: `../img/avatar-angelina.jpg`,
    },
    description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
    location: {
      latitude: 48.851353,
      longitude: 2.341412,
      zoom: 8
    },
  },

]);
