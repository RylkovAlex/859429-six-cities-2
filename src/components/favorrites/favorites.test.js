import React from 'react';
import renderer from 'react-test-renderer';
import {MemoryRouter} from 'react-router';
import {Favorites} from './favorites';
import {favoritesMock} from '../../mocks/offers';
import {citiesMock} from '../../mocks/cities';

jest.mock(`../header/header.jsx`, () => () => <div/>);
jest.mock(`../favorites-empty/favorites-empty.jsx`, () => () => <div/>);
jest.mock(`../offers-list/offers-list.jsx`, () => ({
  __esModule: true,
  default: jest.fn().mockReturnValue(<div/>),
  ListType: {
    MainList: `main`,
    NearbyList: `nearby`,
    FavoriteList: `favorite`,
  }
}));

it(`Favorites correctly renders after relaunch`, () => {
  const props = {
    favorites: favoritesMock,
    postFavorite: jest.fn(),
    activeCity: citiesMock[0],
    cities: citiesMock,
    setCity: jest.fn(),
  };
  const tree = renderer
    .create(
        <MemoryRouter>
          <Favorites {...props}/>
        </MemoryRouter>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
