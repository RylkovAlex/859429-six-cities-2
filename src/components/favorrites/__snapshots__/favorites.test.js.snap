// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites correctly renders after relaunch 1`] = `
<div
  className="page"
>
  <div />
  <main
    className="page__main page__main--favorites"
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites"
      >
        <h1
          className="favorites__title"
        >
          Saved listing
        </h1>
        <ul
          className="favorites__list"
        >
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations "
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="/"
                  onClick={[Function]}
                >
                  <span>
                    Paris
                  </span>
                </a>
              </div>
            </div>
            <div />
          </li>
        </ul>
      </section>
    </div>
  </main>
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="main.html"
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="img/logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;
